ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp

```diff
@@ -352,50 +352,60 @@ VirtualizationRootResult VirtualizationRoot_RegisterProviderForPath(PrjFSProvide
+                            root.providerPid = clientPID;
+                            virtualizationRootVNode = NULLVP; // transfer ownership
+                        }
                     }
                     else
                     {
-                        VirtualizationRoot& root = s_virtualizationRoots[rootIndex];
-                        root.providerUserClient = userClient;
-                        root.providerPid = clientPID;
-                        virtualizationRootVNode = NULLVP; // transfer ownership
+                        rootIndex = InsertVirtualizationRoot_Locked(userClient, clientPID, virtualizationRootVNode, rootVid, vnodeIds, virtualizationRootCanonicalPath);
+                        if (rootIndex >= 0)
+                        {
+                            assert(rootIndex < s_maxVirtualizationRoots);
+                        
+                            virtualizationRootVNode = NULLVP; // prevent vnode_put later; active provider should hold vnode reference
+                        
+                            KextLog_Note("VirtualizationRoot_RegisterProviderForPath: new root not found in offline roots, inserted as new root with index %d. path '%s'", rootIndex, virtualizationRootPath);
+                        }
+                        else
+                        {
+                            // TODO: scan the array for roots on mounts which have disappeared, or grow the array
+                            KextLog_Error("VirtualizationRoot_RegisterProviderForPath: failed to insert new root");
+                        }
                     }
-                }
-                else
-                {
-                    rootIndex = InsertVirtualizationRoot_Locked(userClient, clientPID, virtualizationRootVNode, rootVid, vnodeIds, virtualizationRootPath);
-                    if (rootIndex >= 0)
-                    {
-                        assert(rootIndex < s_maxVirtualizationRoots);
-                        VirtualizationRoot* root = &s_virtualizationRoots[rootIndex];
-                    
-                        strlcpy(root->path, virtualizationRootPath, sizeof(root->path));
-                        virtualizationRootVNode = NULLVP; // prevent vnode_put later; active provider should hold vnode reference
                     
-                        KextLog_Note("VirtualizationRoot_RegisterProviderForPath: new root not found in offline roots, inserted as new root with index %d. path '%s'", rootIndex, virtualizationRootPath);
-                    }
-                    else
+                    if (0 == err)
                     {
-                        // TODO: scan the array for roots on mounts which have disappeared, or grow the array
-                        KextLog_Error("VirtualizationRoot_RegisterProviderForPath: failed to insert new root");
+                        userClient->setProperty(PrjFSProviderPathKey, virtualizationRootCanonicalPath);
```

Comment:
What is the purpose of setting this properly on `userClient`, is it used anywhere?

Reply:
It's a diagnostic aid - you can look in the IO Registry using `ioreg` or IORegistryExplorer and check the path associated with each provider. (This is explained in the commit message IIRC.) I suppose you could argue we should only set it in debug builds, but I don't see any harm, and the updates are infrequent, so not a performance concern.